tsdb/engine/tsm1/cache.go

```diff
@@ -81,6 +81,7 @@ const (
 type Cache struct {
 	mu      sync.RWMutex
 	store   map[string]*entry
+	dirty   map[string]*entry
```

Comment:
@jonseymour My point is that `dirty` is serving exactly the same purpose as what the `store` of the snapshot cache was intended to serve.  Adding another map makes the cache more complicated (3 maps instead of 2) which may solve the race, but it is adding another layer onto something that is still broken IMO.  

The problem I see is that the encapsulation of `entry`, `snapshot`, etc.. is violated without adequate protection.  For example, `entry`'s private state is accessed concurrently, but it doesn't have a mutex protecting it when it is accessed concurrently.  The snapshot's `store` was supposed to be safe copy of the cache to eliminate any chance of concurrent access which is what I believe `dirty` is also intended to do as well.  Seems like the snapshot's `store` should be make a full safe, copy or the unsafe accesses in `entry`, etc.. should be protected.